ln-gateway/src/messaging.rs

```diff
@@ -0,0 +1,48 @@
+use std::sync::Arc;
+
+use anyhow::Error;
+use async_trait::async_trait;
+use tokio::sync::{mpsc, oneshot, Mutex};
+
+use crate::{GatewayRequest, GatewayRequestTrait, Result};
+
+#[async_trait]
+pub trait GatewayMessageReceiver: Send + Sync + 'static {
+    async fn receive(&self) -> ();
+}
+
+#[derive(Debug, Clone)]
+pub struct GatewayMessageSender {
+    sender: Arc<Mutex<mpsc::Sender<GatewayRequest>>>,
+}
+
+/// A two-way message sender for [`GatewayRequest`]s.
+impl GatewayMessageSender {
+    pub fn new(sender: &mpsc::Sender<GatewayRequest>) -> Self {
+        Self {
+            sender: Arc::new(Mutex::new(sender.clone())),
+        }
+    }
+
+    /// Sends a gateway request as message to some gateway receiver over an mpsc channel.
+    ///     in doing so, a oneshot channel is created to receive a response
+    ///     from the receiver of original message
+    ///
+    /// ### Returns
+    ///
+    /// * `Result` - `Ok(Response)` if the message was sent successfully and we got a response, `Err(anyhow:Error)` otherwise
+    pub async fn send<R>(&self, message: R) -> std::result::Result<R::Response, Error>
+    where
+        R: GatewayRequestTrait,
+    {
+        let (sender, receiver) = oneshot::channel::<Result<R::Response>>();
+        let msg = message.to_enum(sender);
+
+        let gw_sender = { self.sender.lock().await.clone() };
+        gw_sender
+            .send(msg)
+            .await
+            .expect("failed to send over channel");
+        Ok(receiver.await.expect("Failed to send over channel")?)
+    }
+}
```

Comment:
`send` is a slightly odd name for this method because usually "sending" on a "receiver" is "fire and forget". But here we await a response. For this reason I called the previous version of this `gw_rpc` to indicate the "send and receive" nature of it.

Not sure how to improve, but just pointing out something I noticed ...